File: python_leaoai/services/ApplicationServices/Core/GenericService.py
from typing import TypeVar, Generic, List
from sqlalchemy.orm import Session
from services import ItemService
from repositories import Repository
from exceptions import NotFoundException, InvalidArgumentException

T = TypeVar("T")  # Tipo de entidade

class GenericService(Generic[T]):
    def __init__(self, db: Session, item_service: ItemService, repository: Repository[T]):
        self._db = db
        self._srvItem = item_service
        self._repository = repository

    def list_item(self) -> List[dict]:
        try:
            items = self._repository.get_all_ativo()
            return [self._srvItem.build(item) for item in items]
        except Exception as ex:
            raise Exception("An error occurred while retrieving items") from ex

    def get_item_by_id(self, id: int) -> dict:
        if id <= 0:
            raise InvalidArgumentException("ID must be greater than zero")

        try:
            item = self._repository.get_obj(id)
            if not item:
                raise NotFoundException(f"Item with ID {id} not found")
            return self._srvItem.build(item)
        except Exception as ex:
            raise Exception("An error occurred while retrieving the item by ID") from ex

    def get_item_by_url(self, url: str) -> dict:
        if not url:
            raise InvalidArgumentException("URL cannot be null or empty")

        try:
            item = self._repository.get_by_url(url)
            if not item:
                raise NotFoundException(f"Item with URL '{url}' not found")
            return self._srvItem.build(item)
        except Exception as ex:
            raise Exception("An error occurred while retrieving the item by URL") from ex

File: Framework.Services/ApplicationServices/Core/GenericService.cs
using Data;
using Framework.Domain.Site.Component;
using Framework.Repositories;
using Framework.Services.Site.Interfaces;
using System.Collections.Generic;
using System.Linq;

namespace Framework.Services.ApplicationServices.Core
{
    public class GenericService<T, R> : IService where R : IRepository<T> where T : class
    {
        private readonly ApplicationDbContext _context;
        private readonly ItemService _srvItem;
        private readonly R _repository;

        public GenericService(ApplicationDbContext context, ItemService itemService, R repository)
        {
            _context = context;
            _srvItem = itemService;
            _repository = repository;
        }

        public List<Item> ListItem()
        {
            try
            {
                var items = _repository.GetAllAtivo();
                return items.Select(item => _srvItem.Build(item)).ToList();
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while retrieving items", ex);
            }
        }

        public Item GetItemById(int id)
        {
            try
            {
                if (id <= 0)
                {
                    throw new ArgumentException("ID must be greater than zero", nameof(id));
                }

                var item = _repository.GetObj(id);
                return _srvItem.Build(item);
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while retrieving the item by ID", ex);
            }
        }

        public Item GetItemByUrl(string url)
        {
            try
            {
                if (string.IsNullOrEmpty(url))
                {
                    throw new ArgumentException("URL cannot be null or empty", nameof(url));
                }

                var item = _repository.GetByUrl(url);

                return _srvItem.Build(item);
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while retrieving the item by URL", ex);
            }
        }

    }
}
